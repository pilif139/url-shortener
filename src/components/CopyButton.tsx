'use client'

import {useState} from "react";

export default function CopyButton({text}: {text: string}) {
  const [isCopied, setIsCopied] = useState(false);
  const copyToClipboard = async (text: string) => {
    try {
      await navigator.clipboard.writeText(text);
      setIsCopied(true);
    } catch (error) {
      console.error("Failed to copy text", error);
    }
  }

  return (
      <button onClick={() => copyToClipboard(text)}
              className="bg-violet-500 text-xl p-2 rounded-xl hover:bg-violet-600 text-white">
        {isCopied ? "Copied!" : "Copy"}
      </button>
  );
}