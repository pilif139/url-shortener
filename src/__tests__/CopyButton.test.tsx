import CopyButton from "@/components/CopyButton";
import {render, screen, waitFor} from "@testing-library/react";
import {jest} from "@jest/globals";
import {act} from "react";

describe("CopyButton", () => {
  beforeEach(() => {
    Object.defineProperty(navigator, "clipboard", {
      value: {
        writeText: async (text: string) => {
          expect(text).toBe("Hello, World!");
        },
        readText: async () => {
          return "Hello, World!";
        }
      },
      configurable: true
    });
  });

  it("should render CopyButton", () => {
    const {container} = render(<CopyButton text="Hello, World!" />);
    expect(container).toMatchSnapshot();
  });

  it("should render Copied! when clicked", async () => {
    render(<CopyButton text="Hello, World!" />);
    const button = screen.getByRole("button");
    expect(button).toBeInTheDocument();
    expect(screen.getByText("Copy")).toBeInTheDocument()
    act(() => {
      button.click();
    });
    await waitFor(()=>{
      expect(screen.getByText("Copied!")).toBeInTheDocument()
    });
  });

  it("should copy text to clipboard", async() => {
    render(<CopyButton text="Hello, World!" />);
    await act(async()=>{
      screen.getByText("Copy").click();
      expect(await navigator.clipboard.readText()).toBe("Hello, World!");
    })
  });

  it("should log error when failed to copy text", async () => {
    const consoleSpy = jest.spyOn(console, "error");
    consoleSpy.mockImplementation(() => {});
    render(<CopyButton text="Hello, World!" />);
    Object.defineProperty(navigator, "clipboard", {
      value: {
        writeText: async () => {
            throw new Error("Failed to copy text");
        }
      }
    });
    act(()=> {
      screen.getByText("Copy").click();
    });
    await waitFor(()=>expect(consoleSpy).toHaveBeenCalled());
    consoleSpy.mockRestore();
  });
});