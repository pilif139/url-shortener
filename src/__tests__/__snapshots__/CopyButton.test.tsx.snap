// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CopyButton should render CopyButton 1`] = `
<div>
  <button
    class="rounded-xl bg-violet-500 p-2 text-xl text-white hover:bg-violet-600"
  >
    Copy
  </button>
</div>
`;
